client/my-sites/checkout/checkout-thank-you/jetpack-instruction-list.tsx

```diff
@@ -0,0 +1,19 @@
+import React, { ReactNode } from 'react';
+
+export interface JetpackInstructionListProps {
+	items: ReactNode[];
+}
+
+const JetpackInstructionList: React.FC< JetpackInstructionListProps > = ( { items } ) => {
+	return (
+		<ol className="jetpack-instruction-list">
+			{ items.map( ( item, index ) => (
+				<li className="jetpack-instruction-list__item" key={ index }>
```

Comment:
Please don't use an index as a React key.

Using an index as a React key is an easy way to remove the warning that shows up in the browser when no key is provided, but it in fact is a [really](https://reactpatterns.js.org/docs/indexes-as-a-key-is-an-anti-pattern/) [dangerous](https://articles.wesionary.team/using-index-as-a-key-is-an-anti-pattern-in-react-8e5db3aea3a9) [thing](https://robinpokorny.com/blog/index-as-a-key-is-an-anti-pattern/) [to](https://dev.to/shiv1998/why-not-to-use-index-as-key-in-react-lists-practical-example-3e66) [do](https://www.youtube.com/watch?v=xlPxnc5uUPQ).

The key prop is used by React to uniquely identify each component and match it with the part of the DOM (the rendered web page) it represents; this way React can update the DOM when its components (the virtual DOM) change over time. If you are rendering a list of elements keyed by their index (eg: the components A, B, C), and a new element is added in the middle (eg: A, X, B, C), the newly-added component will suddenly have the index of a component that already existed (eg: the component X will have index 1 which previously belonged to component B) . React will think that component is unchanged and may not render it at all, or may render something very unexpected.

This gets worse. In the above example, if the user goes to edit or delete component B, they may actually edit component X! This can cause all kinds of problems. It's best to come up with some data which is unique and use that as a key instead.

In this case, since the items are effectively any React nodes, it might be tricky to do. 🤔  I suggest looking up how React itself deals with `children` props, which is similar to what you're doing here. You are just wrapping each child in an `li` and a `span`. If you can't find a way to uniquely identify the nodes that way, you might want to consider one of two options:

1. Change this component to replace `{ items.map...}` with `{ children }` and then change how this is rendered so the children are included directly in the parent (inversion of control pattern).
2. Change the type of `items` to something like `items: { id: string; item: ReactNode }[]` so that you can include a unique string for each item in the parent.

Reply:
Tangential:

AFAIK, index-as-key is pretty pervasive inside Calypso. Considering the danger, @sirbrillig do you know if the Calypso team is aware of this issue?

Reply:
Good question. I'm pretty sure they know about the anti-pattern as it's so widely documented (and has been for the life of React). The reason it's probably gotten a pass thus far is that, in many cases, the data being rendered by a `map()` statement like this never changes (it's hard-coded into the parent). In that case, `key={index}` _is actually safe_, but IMO it's better to just not add a `key` at all in those cases, since `index` is actually the default if it's not set. That will generate a console warning, but it's an accurate one. 

Typically I believe it's better to just create a unique key, though, because that way if the parent of a component ever changes how it provides the data, you won't end up with suddenly appearing (and difficult to find) bugs. Effectively using it is a [foot-gun](https://en.wiktionary.org/wiki/footgun) waiting for some future developer to pull the trigger.

In this case, my second suggestion might be the easiest one. We could change this:

```js
const items = useMemo(
		() => [
			translate( 'Go to your WP Admin Dashboard and {{strong}}add a new plugin{{/strong}}.', {
				components: { strong: <strong /> },
			} ),
			// ...
		],
		[ translate ]
	);
```

to this:

```js
const items = useMemo(
		() => [
			{ id: 'go-to-wp-admin', item: translate( 'Go to your WP Admin Dashboard and {{strong}}add a new plugin{{/strong}}.', {
				components: { strong: <strong /> },
			} ) },
			// ...
		[ translate ]
	);
```

Reply:
If the items _within_ the list do not change, it's perfectly fine to use the index as the key. A dedicated unique key is important when user interaction results in items inside the list being added/removed/updated. So, in this case here, the list is not going to change, so it's fine to use index here.

Reply:
Dan Abramov
https://twitter.com/dan_abramov/status/1415283502124261378

Reply:
I do agree with the approach of passing an explicit `id` as the key.

Reply:
@sirbrillig @manzoorwanijk While it is true that the current component is used for a static list and that `{key= index }`would suffice in this case, There is still a possibility that the component will be used with dynamic list in the future if someone decides to use the same component. The second approach by @sirbrillig is good and easy to follow. 👍